### PR TITLE
Metadata query language

### PR DESCRIPTION
**Pull request by BeckySharp**
@marcovzla  and I worked on a query language for the metadata.
This supports:
- Dates
- Numbers
- strings
- regex over strings
- nested fields (e.g., people, where you want to match against attributes of a metadata field without mixing)

Other:
- tests added and older tests related to metadata queries updated.
- made slight change from #281 in terms of getting the field name and type for the metadata docs -- getting from a constant variable in `OdinsonIndexWriter` rather than the config for simplicity.
- added access methods in `ExtractorEngine` for `mkQuery` and `mkFilteredQuery` (i.e., the query + parent/metadata query), updated the tests to use this as example to outside world.

### CODE REVIEW

File: backend/test/controllers/OdinsonControllerSpec.scala [hunk 1]
```diff
@@ -148,7 +148,7 @@ class OdinsonControllerSpec extends PlaySpec with GuiceOneAppPerTest with Inject
       val res2 = controller.runQuery(
         odinsonQuery = "[lemma=pie]",
         // See http://www.lucenetutorial.com/lucene-query-syntax.html
```
Comment (myedibleenso) [inline backend/test/controllers/OdinsonControllerSpec.scala:150]: We should remove this comment, as it's no longer relevant.

**Review by myedibleenso — COMMENTED** [review]
Great!  

The tests help illustrate features, but I think this calls for revisions to our documentation before merging.  We should provide a small overview of the language and correct prior mentions that claim the parent/metadata query is a Lucene query and instructions for preparing the metadata in the Odinson Document (ex. are there restrictions on field types?).

We also should change the API spec (a yaml file in the backend project), routes file, and param names for the corresponding method(s).  

Are we going to call this a metadata query now?  If so, I'd like to change some of the documentation in the REST API to reflect that convention.

**Review by danebell — CHANGES_REQUESTED** [review]
My only request is that the key `parentDocFieldFileName` be available in `reference.conf`. Otherwise, this looks great.

**Comment by danebell** [comment]
Merging in tests that rely on [`ExtractorEngine.getParentDoc`](https://github.com/lum-ai/odinson/blob/metadata/core/src/main/scala/ai/lum/odinson/ExtractorEngine.scala#L214-L228), it seems that the changes to `getParentDoc` make it crash because no documents ever match. I believe this is because it anticipates that the document ID will be part of the metadata, when in practice, it's outside the metadata group in the JSON.

File: core/src/main/resources/reference.conf [hunk 2]
```diff
@@ -94,13 +94,6 @@ odinson {
         ${odinson.index.wordTokenField},
     ]
 
-    # fields for referencing the parent doc
-    parentDocField = metadata
-
-    parentDocFieldType = type
-
-    parentDocFieldFileName = fileName
```
[the remark this answers is not in the document]
Comment (BeckySharp) [reply core/src/main/resources/reference.conf:106]: addressed with #304, right?

**Comment by BeckySharp** [comment]
> Great!
> 
> The tests help illustrate features, but I think this calls for revisions to our documentation before merging. We should provide a small overview of the language and correct prior mentions that claim the parent/metadata query is a Lucene query and instructions for preparing the metadata in the Odinson Document (ex. are there restrictions on field types?).

@myedibleenso , we added documentation for the Metadata queries (`metadata.md`).

> We also should change the API spec (a yaml file in the backend project), routes file, and param names for the corresponding method(s).

Are we to do this? or will you/ @danebell ?

> Are we going to call this a metadata query now? If so, I'd like to change some of the documentation in the REST API to reflect that convention.

Yes, I think we should call them metadata queries, as it's more true to their semantics, `parent` is more about the implementation in lucene.

**Comment by myedibleenso** [comment]
@danebell , can you take a look at this and take a first pass on updating the backend (API, routes file, tests, and OpenAPI spec)?  We should probably replace mentions of `parentQuery` with`metadataQuery`.

File: core/src/main/resources/reference.conf (hunk 2 above)
[the remark this answers is not in the document]
Comment (danebell) [reply core/src/main/resources/reference.conf:106]: The fieldname is now available, but as referenced in my [other comment](https://github.com/lum-ai/odinson/pull/305#issuecomment-870866868), `getParentDoc` is not working as expected. If I see an obvious solution while working on #312 I'll apply it.

**Pull request merged**